library/core/src/ptr/mod.rs

```diff
@@ -63,12 +63,212 @@
+//! and then immediately converting back to a pointer. To make this use case more ergonomic,
+//! we provide the [`map_addr`][] method.
+//!
+//! To help make it clear that code is "following" Strict Provenance semantics, we also
+//! provide an [`addr`][] method which is currently equivalent to `ptr as usize`. In the
+//! future we may provide a lint for pointer<->integer casts to help you audit if your
+//! code conforms to strict provenance.
+//!
+//!
+//! ## Using Strict Provenance
+//!
+//! Most code needs no changes to conform to strict provenance, as the only really concerning
+//! operation that *wasn't* obviously already Undefined Behaviour is casts from usize to a
+//! pointer. For code which *does* cast a usize to a pointer, the scope of the change depends
+//! on exactly what you're doing.
+//!
+//! In general you just need to make sure that if you want to convert a usize address to a
+//! pointer and then use that pointer to read/write memory, you need to keep around a pointer
+//! that has sufficient provenance to perform that read/write itself. In this way all of your
+//! casts from an address to a pointer are essentially just applying offsets/indexing.
+//!
+//! This is generally trivial to do for simple cases like tagged pointers *as long as you
+//! represent the tagged pointer as an actual pointer and not a usize*. For instance:
+//!
+//! ```
+//! #![feature(strict_provenance)]
+//!
+//! unsafe {
+//!     // A flag we want to pack into our pointer
+//!     static HAS_DATA: usize = 0x1;
+//!     static FLAG_MASK: usize = !HAS_DATA;
+//!
+//!     // Our value, which must have enough alignment to have spare least-significant-bits.
+//!     let my_precious_data: u32 = 17;
+//!     assert!(core::mem::align_of::<u32>() > 1);
+//!
+//!     // Create a tagged pointer
+//!     let ptr = &my_precious_data as *const u32;
+//!     let tagged = ptr.map_addr(|addr| addr | HAS_DATA);
+//!
+//!     // Check the flag:
+//!     if tagged.addr() & HAS_DATA != 0 {
+//!         // Untag and read the pointer
+//!         let data = *tagged.map_addr(|addr| addr & FLAG_MASK);
+//!         assert_eq!(data, 17);
+//!     } else {
+//!         unreachable!()
+//!     }
+//! }
+//! ```
+//!
+//! Something more complicated and just generally *evil* like a XOR-List requires more significant
+//! changes like allocating all nodes in a pre-allocated Vec or Arena and using a pointer
+//! to the whole allocation to reconstitute the XORed addresses.
+//!
+//! Situations where a valid pointer *must* be created from just an address, such as baremetal code
+//! accessing a memory-mapped interface at a fixed address, are an open question on how to support.
+//! These situations *will* still be allowed, but we might require some kind of "I know what I'm
+//! doing" annotation to explain the situation to the compiler. Because those situations require
+//! `volatile` accesses anyway, it should be possible to carve out exceptions for them.
```

Comment:
I don't think they always require volatile accesses.

Also it's not really an open question, is it? Int-to-ptr casts are fine if they are interpreted as "this generates a fresh provenance and does not interfere with any other provenance generated anywhere else". I.e., if we think of this as `malloc` and "making a noalias promise", all is fine. It's only when you expect the cast to pick up some *previously existing provenance* (i.e. when you round-trip ptr-int-ptr) that we have a problem.

Reply:
MMIO effectively is never used in a `noalias` fashion in actual practice, it's always `*mut T` not `&mut T`.

Reply:
`0xCOFE as *mut T` is still a "noalias" promise in the sense that it must not alias with any pointer returned by an Abstract Machine allocation operation (local variables, statics, heap allocations).

Reply:
So we need something like `with_new_provenance(addr: usize) -> *const T`? It looks less risky a d more obvious than `addr as *const T`, though this is not necessarily have to be included in the MVP.

Reply:
(I still think this text should be adjusted, @Gankra.)

Reply:
> It's possible they need no special attention at all, because they're generally accessing memory outside the scope of "the abstract machine", or already using "I know what I'm doing" annotations like "volatile".

How's that?

Reply:
(I would like to avoid immediately blessing the "from-thin-air == fresh-valid-provenance" interpretation in these docs because it seems hostile to CHERI compatibility.)

Reply:
(Yes I know ZST hax already implies it but... enghhgh....)

Reply:
Ah but you said "or" so this is covered by the first alternative. Okay sounds good to me. :)

Reply:
There's certainly situations where you have a fixed address for something mmio/volatile, and other cases where you have a fixed address for "normal memory" that is simply disjoint from any rust memory.
* In the mmio case, using a type or types to help keep track of the volatile access is a very wise plan (which is why I made it a crate).
* In the disjoint-normal-memory case, modeling this as "a weird allocator" that claims a memory region seems sufficient. This would, as usual, be unsafe, and only really work if you're the CPU kernel.